src/golibmc.go

```diff
@@ -921,13 +1284,50 @@ func (client *Client) Stats() (map[string](map[string]string), error) {
 
 // Quit will close the sockets to each memcached server
 func (client *Client) Quit() error {
-	client.lock()
-	defer client.unlock()
+	client.lk.Lock()
 
-	errCode := C.client_quit(client._imp)
-	C.client_destroy_broadcast_result(client._imp)
+	close(client.openerCh)
+	if client.cleanerCh != nil {
+		close(client.cleanerCh)
+	}
+	for _, cr := range client.connRequests {
+		close(cr)
+	}
+	client.closed = true
+	client.lk.Unlock()
+	for _, cn := range client.freeConns {
+		if err := cn.quit(); err != nil {
+			return err
+		}
+	}
+	client.lk.Lock()
+	client.freeConns = nil
+	client.lk.Unlock()
 
+	return nil
+}
+
+func (cn *conn) expired(timeout time.Duration) bool {
+	if timeout <= 0 {
+		return false
+	}
+	return cn.createdAt.Add(timeout).Before(time.Now())
+}
+
+func (cn *conn) quit() error {
+	cn.Lock()
+	defer cn.Unlock()
+	if cn.closed {
+		return errors.New("duplicate client close")
+	}
+	cn.closed = true
+	errCode := C.client_quit(cn._imp)
+	C.client_destroy_broadcast_result(cn._imp)
 	if errCode == C.RET_CONN_POLL_ERR || errCode == C.RET_RECV_ERR {
+		cn.client.lk.Lock()
+		defer cn.client.lk.Unlock()
+		cn.client.numOpen--
+		cn.client.maybeOpenNewConnections()
```

Comment:
maybeOpenNewConnections here is not Appropriate, for example, connectionCleaner also use quit

https://github.com/douban/libmc/pull/57/files#diff-b21c9a32b7563e26c49d7f215bc5a656R543

Reply:
one more example , Quit will call quit （https://github.com/douban/libmc/pull/57/files#diff-b21c9a32b7563e26c49d7f215bc5a656R1299 ），so may maybeOpenNewConnections 
 here will try to write to closed client.openerCh。

Reply:
sorry for the second example, maybeOpenNewConnections will return immediately  if client.closed is true。

Reply:
okay, maybeOpenNewConnections here seems good to me。

but i wonder the cn.client.numOpen-- here seems not，because if only "errCode == C.RET_CONN_POLL_ERR || errCode == C.RET_RECV_ERR"  then "cn.client.numOpen--", how about quit when badConn happened? if numOpen higher than it should be, will prevent new conn to open when numOpen reached maxOpen？

Reply:
What you want to say is that errCode "C.RET_CONN_POLL_ERR" or "C.RET_RECV_ERR" is not sufficient as a condition to subtract numOpen?
Is there errCode inside which it can be determined that conn is badConn?
https://github.com/douban/libmc/blob/v1.1.0/src/golibmc.go#L42-L52

Reply:
> What you want to say is that errCode "C.RET_CONN_POLL_ERR" or "C.RET_RECV_ERR" is not sufficient as a condition to subtract numOpen?

yes，that's what i want to say :)

i think that 

- RET_MC_SERVER_ERR https://github.com/douban/libmc/blob/7d6700d05b115208b5bce676e272c70d4e4a7645/src/ConnectionPool.cpp#L374 

- RET_SEND_ERR https://github.com/douban/libmc/blob/7d6700d05b115208b5bce676e272c70d4e4a7645/src/ConnectionPool.cpp#L432 

also need to be considered to subtract numOpen in addition to  "C.RET_CONN_POLL_ERR" or "C.RET_RECV_ERR"。

Reply:
Thanks for your explain.
Fixed, do you agree with what you are imagining?
0a37f19...10a042a